typed-racket-lib/typed-racket/private/type-contract.rkt

```diff
@@ -361,6 +361,12 @@
          (if numeric-sc
              (apply or/sc numeric-sc (map t->sc non-numeric))
              (apply or/sc (map t->sc elems)))]
+        [(and t (Function: arrs))
+         #:when (any->bool? arrs)
+         (or/sc (and/sc
+                  (flat/sc #'struct-predicate-procedure?)
+                  (flat/sc #'(compose1 not chaperone?)))
```

Comment:
Don't use `compose` when you care about performance, unless you're sure it's inlined.